### PR TITLE
[#2108] render attachments via whatsapp

### LINKED ISSUE #2108: Render/Send File Attachments via WhatsApp
We need to be able to render image, video, text, and PDF files via WhatsApp. We should also be able to send them via WhatsApp through in the UI.

- Image, video, text and PDF files up to 5mb in size
- Can send media messages up to 16mb in size (including the image, video, text, or PDF file.
- To send a media message with WhatsApp, we need to include the `MediaUrl `parameter in your `HTTP POST requests`

Media content is currently only supported in Session Messages. Messaging sessions are created when a user responds to a template message you send them, or starts a conversation by messaging your number on WhatsApp. These sessions stay active for 24 hours after the last message they send you. If the WhatsApp session with a user expires, you must wait for an inbound message to create a new session before you can send them a media message.

Through Twilio:WhatsApp, the following file types are possible:
- Images -- JPG, JPEG, PNG
- Audio  -- MP3, OGG, AMR
- Documents -- PDF
- Video -- MP4 (with H.264 video codec and AAC audio)

### Design Notes
122225089-bf91b600-ce82-11eb-9a1b-012bfacda270

We can use the same order of the icons within the input bar that we already have (emojis, templates, attachments).

### LINKED ISSUE #2110: Render/Send File Attachments via Google Business Messenger
We need to offer the ability to render/send file attachments via GBM.

Users can attach files up to 200 MB from their computer, mobile device, or Google Drive directly to Google Chat messages.

### Common file types
By default, all common file types are matched by file format and filename to identify the type of content and the attachment's extension. Examples include image/photo.jpg or document/letter.doc.

The most common supported image files:
- BMP
- GIF
- JPEG
- PNG
- WBMP
- HEIC

### Custom file types
By default, custom file types are matched only by filename, for example, photo.jpg or letter.doc.

Note: Generally speaking, most formats get converted into text or html (including unknown formats). Content matches are then made against the resultant normalized data. 

### Supported Document Types
.cpr, .cwk, .cws, .dcx, .doc, .docm, .docx, .dot, .dotm, .dotx, .fax, .fp, .fp3, .frm, .gim, .gix, .gna, .gnx, .gra, .mcw, .mdb, .mdn, .met, .mpp, .obd, .odg, .odp, .ods, .odt, .pdf, .potm, .potx, .ppam, .pps, .ppsm, .ppsx, .ppt, .pptm, .pptx, .pre, .prs, .rtf, .shb, .shw, .sldx, .sldm, .wb1, .wb2, .wdb, .wk1, .wk3, .wk4, .wks, .wp, .wpw, .wp4, .wp5, .wp6, .wpd, .wps, .wpt, .wq1, .wq2, .wri, .ws1, .ws2, .ws3, .ws4, .ws5, .ws6, .ws7, .wsd, .xlam, .xlm, .xls, .xlsb, .xlsm, .xlsx, .xlt, .xltm, .xltx

### Supported Image Types
.ai, .art, .att, .bmp, .cal, .cdr, .cdt, .cdx, .cmf, .cmp, .dib, .drw, .emf, .eps, .fh3, .fif, .fpx, .gem, .gif, .html, .icb, .iff, .ima, .img, .jbf, .jff, .jif, .jpeg, .jpg, .jtf, .kdc, .kfx, .lbm, .mac, .mic, .pbm, .pcd, .pcs, .pct, .pcx, .pgm, .pic, .pif, .png, .pnt, .ppm, .ps, .psd, .ras, .raw, .sct, .sdr, .sdt, .sep, .shg, .tga, .tif, .tiff, .vda, .vst, .wil, .wmf, .wpg, .wvl

### Supported Audio (music + sound) Types
.aif, .aiff, .ams, asf, .cda, .dcr, .dsm, .idd, .it, .mdl, .med, .mid, .mp3, .mtm, .mod, .mus, .nsa, .ra, .ram, .rm, .rmi, .rtm, .snd, .stm, .svx, .s3m, .ult, .voc, .wav, .wow

### Supported Video + Multimedia Types
.avi, .cfb, .cmv, .dir, .gal, .m3d, .mmm, .mov, .mpe, .mpeg, .mvb, .qt, .qtm, .xtp, .xy3, .xy4, .xyp, .xyw, .mpg, .wmv

**Note**: With video files, we should not have a video player in the beginning. Instead, we should allow users to download the file and then watch it.

### Compressed and archive file formats
.7z, .ace, .bz, .bz2, .cab, .gz, .hex, .hqx, .lzh, .pages, .rar, .sea, .sit, .tar, .tgz, .uue, .zip, .zoo

### Custom Types
asc., .brf, .c, .c++, .cc, .cpp, .css, .csv, .cxx, .eps2, .eps3, .epsf, .epsi, .epub, .es, .fnt, .fon, .h, .h++, .hh, .hpp, .hwp, .hxx, .jpe, .js, .kml, .kmz, .markdown, .md, .mdown, .mhtml, .mkd, .mkdn, .numbers, .otf, .otp, .ott, .php, .pht, .phtml, .pot, .qjs, .qjsp, .sdc, .sdd, .sdw, .shtm, .shtml, .svg, .svgz, .swf, .swfl, .sxc, .sxd, .sxi, .sxw, .tsv, .ttf, .wml, .xhtml, .xlb, .xml, .xps, .xsd, .xsl
Note: Although you can enter unsupported custom file types, they're matched only by filename, for example, document.sac.

### Blocked File Types
You can’t attach certain types of files. These include:
ADE, ADP, APK, BAT, CAB, CHM, CMD, COM, CPL, DLL, DMG, EXE, HTA, INS, ISP, JAR, JS, JSE, LIB, LNK, MDE, MSC, MSI, MSP, MST, NSH, PIF, SCR, SCT, SHB, SYS, VB, VBE, VBS, VXD, WSC, WSF, and WSH.
This list includes any compressed form, such as GZ or BZ2 files, or files found in archives, such as ZIP or TGZ files.
Documents with malicious macros.
Password-protected archives whose content is an archive.

### Design Note

In terms of design for the UI, we can use the image below as a reference.

122225089-bf91b600-ce82-11eb-9a1b-012bfacda270

We can use the same order of the icons within the input bar that we already have (emojis, templates, attachments).

### Design Note
Please use the paperclip icon from [this website](https://www.zwicon.com/cheatsheet.html):
Screenshot 2021-06-16 at 08 56 56

### LINKED ISSUE #2571: Improve InputSelector
- improve scaling down 
- improve scaling of very big / very small images

### PR DESCRIPTION
**Pull request by AudreyKj**
closes #2108 
closes #2571
closes #2110

- improved inputSelector 
- improved sending attachments experience 
- refactored media attachment code 
- added WA rendering + sending 
- added GBM sending (as we already render all supported messages)

### CODE REVIEW

File: frontend/ui/src/pages/Inbox/MessageInput/InputOptions.tsx [hunk 1]
```diff
@@ -147,28 +154,30 @@ export const InputOptions = (props: Props) => {
...
+            </label>
+
+            <input
+              type="file"
+              id="file"
+              name="file"
+              onChange={selectFile}
+              className={styles.fileInput}
+              disabled={inputDisabled || loadingSelector}
+              accept={inputAcceptedFiles}
```
Comment (AudreyKj) [inline frontend/ui/src/pages/Inbox/MessageInput/InputOptions.tsx:177]: A list of accepted files is dynamically added for each source: unsupported files cannot be uploaded.

File: frontend/ui/src/pages/Inbox/MessageInput/InputSelector.module.scss [hunk 2]
```diff
@@ -3,7 +3,17 @@
...
+}
+
+@keyframes fadeIn {
+  0% {
+    opacity: 0;
+  }
+  100% {
+    opacity: 1;
+  }
 }
```
Comment (AudreyKj) [inline frontend/ui/src/pages/Inbox/MessageInput/InputSelector.module.scss:17]: fadeIn animation added to `InputSelector`

File: frontend/ui/src/pages/Inbox/MessageInput/InputSelector.tsx [hunk 3]
```diff
@@ -3,71 +3,75 @@ import styles from './InputSelector.module.scss';
...
+      const fileSelectorHeight = entry.contentRect.height;
+      const fileSelectorWidth = entry.contentRect.width;
+
+      if (fileSelectorHeight > contentResizedHeight) {
+        scaleDownInputSelector(fileSelectorHeight);
+      } else if (fileSelectorHeight >= textareaHeight && fileSelectorWidth > minImageHeight) {
+        setSelectorPreviewCloseButton(true);
+      }
+    }
+  });
```
Comment (AudreyKj) [inline frontend/ui/src/pages/Inbox/MessageInput/InputSelector.tsx:38]: refactored InputSelector using `ResizeObserver` to dynamically track DOM size changes

File: frontend/ui/src/pages/Inbox/MessageInput/index.tsx [hunk 4]
```diff
@@ -187,38 +184,27 @@ const MessageInput = (props: Props) => {
...
-      )}, ${fileExtensions.join(', ')}`;
-      return setFileUploadErrorPopUp(message);
+    //unsupported file error
+    if (!getAttachmentType(file.name, source)) {
+      const supportedFilesForSource = getAllSupportedAttachmentsForSource(source);
+
+      const errorMessage = `This file type is not supported by this source. 
+      Supported files: ${supportedFilesForSource}`;
+
+      return setFileUploadErrorPopUp(errorMessage);
```
Comment (AudreyKj) [inline frontend/ui/src/pages/Inbox/MessageInput/index.tsx:205]: dynamically lists supported files for each source

File: frontend/ui/src/services/types/attachmentsTypes.ts [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    audioFiles && docsFiles ? '.' + audioFiles.join(', .') + ',' : audioFiles ? audioFiles.join(', .') : '';
+  const supportedVideoFiles =
+    videoFiles && audioFiles ? '.' + videoFiles.join(', .') + ',' : videoFiles ? videoFiles.join(', .') : '';
+  const supportedImageFiles =
+    imageFiles && videoFiles ? imageFiles.join(', .') + ',' : imageFiles ? imageFiles.join(', .') : '';
+
+  const inputAcceptValue = `.${supportedImageFiles} ${supportedVideoFiles} ${supportedAudioFiles} ${supportedDocsFiles}`;
+
+  return inputAcceptValue;
+};
```
Comment (AudreyKj) [inline frontend/ui/src/services/types/attachmentsTypes.ts:45]: added to `/Services` functions that dynamically fetch that supported files for each source

File: lib/typescript/render/SourceMessagePreview.tsx [hunk 6]
```diff
@@ -37,75 +39,114 @@ export const SourceMessagePreview = (props: SourceMessagePreviewProps) => {
...
 
-    if (lastMessageContent.image) {
+    if (lastMessageContent?.image) {
       return <AttachmentImage />;
     }
 
-    if (lastMessageContent.richCard) {
+    if (lastMessageContent?.richCard) {
       return <RichCardIcon style={{height: '24px', width: '24px', margin: '0px'}} />;
     }
```
Comment (AudreyKj) [inline lib/typescript/render/SourceMessagePreview.tsx:152]: fixed SourceMessagePreview for twilio/WA + added the attachment preview for WA

File: lib/typescript/render/components/Image/index.tsx [hunk 7]
```diff
@@ -1,28 +1,35 @@
...
+            />
+          ))}
+        </div>
+      ) : (
+        <ImageWithFallback className={styles.messageListItemImageBlock} src={imageUrl} alt={altText} />
+      )}
+    </div>
+
+    {text && <Text text={text} fromContact={fromContact} />}
+  </>
```
Comment (AudreyKj) [inline lib/typescript/render/components/Image/index.tsx:34]: added Text to image for WA text captions

File: lib/typescript/render/components/Video/index.tsx [hunk 8]
```diff
@@ -26,19 +29,23 @@ export const Video = ({videoUrl}: VideoRenderProps) => {
...
+            </a>
+          )}
+        </div>
       </div>
-    </div>
+
+      {text && <Text text={text} fromContact={fromContact} />}
+    </>
   );
 };
```
Comment (AudreyKj) [inline lib/typescript/render/components/Video/index.tsx:51]: added Text to video for WA text captions

File: lib/typescript/render/outbound/google.ts [hunk 9]
```diff
@@ -13,4 +13,19 @@ export class GoogleMapper extends OutboundMapper {
...
+      },
+      fallback: 'An image has been sent with Google Business Messages.',
+      image: {
+        contentInfo: {
+          altText: 'An image sent via Google Business Messages.',
+          fileUrl: mediaUrl,
+        },
+      },
+    };
+  }
```
Comment (AudreyKj) [inline lib/typescript/render/outbound/google.ts:30]: attachment for Google

File: lib/typescript/render/outbound/twilio.ts [hunk 10]
```diff
@@ -10,4 +10,10 @@ export class TwilioMapper extends OutboundMapper {
   isTextSupported(): boolean {
     return true;
   }
+
+  getAttachmentPayload(mediaUrl: string): any {
+    return {
+      MediaUrl: mediaUrl,
+    };
+  }
```
Comment (AudreyKj) [inline lib/typescript/render/outbound/twilio.ts:18]: attachment for Twilio

File: lib/typescript/render/providers/twilio/TwilioRender.tsx [hunk 11]
```diff
@@ -13,28 +15,203 @@ function render(content: ContentUnion, props: RenderPropsUnion) {
...
+    }
+  }
+
+  //text
   return {
     type: 'text',
-    text: messageContent.Body,
+    text: messageContent?.Body ?? 'Unsupported message type',
   };
 };
```
Comment (AudreyKj) [inline lib/typescript/render/providers/twilio/TwilioRender.tsx:217]: Twilio rendering for WA

File: lib/typescript/render/providers/twilio/components/CurrentLocation/index.tsx [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+        <Emoji symbol={'📍'} /> This user has shared its current location.
+      </span>
+      <br />
+      <span className={styles.geolocation}>
+        {' '}
+        Latitude: {latitude}, Longitude: {longitude}
+      </span>
+    </div>
+  );
+};
```
Comment (AudreyKj) [inline lib/typescript/render/providers/twilio/components/CurrentLocation/index.tsx:18]: CurrentLocation for WA

File: lib/typescript/render/services/decodeURIComponentMessage.ts [hunk 13]
```diff
@@ -0,0 +1,15 @@
...
+
+  const enCodedMessage = messageContent.substring(
+    enCodedMessageStartIndex + enCodedMessageStartLength,
+    enCodedMessageEndIndex
+  );
+  const formattedEnCodedMessage = enCodedMessage.split('+').join(' ');
+  const decodedMessage = decodeURIComponent(formattedEnCodedMessage);
+
+  return decodedMessage;
+};
```
Comment (AudreyKj) [inline lib/typescript/render/services/decodeURIComponentMessage.ts:15]: abstracted the function that decodes URIComponentMessage in `Services`, as we were using it repeatedly for Twilio

File: lib/typescript/render/outbound/chatplugin.ts [hunk 14]
```diff
@@ -10,4 +11,17 @@ export class ChatpluginMapper extends OutboundMapper {
...
+    return {
+      attachment: {
+        type: mediaType,
+        payload: {
+          url: mediaUrl,
+        },
+      },
+    };
+  }
 }
```
Comment (AudreyKj) [inline lib/typescript/render/outbound/chatplugin.ts:27]: added `getAttachmentPayload` to the Chatplugin outbound mapper to be able to use the OutboundMapper type (but obviously this should be completed with your PR @thorstenairy)

**Review by thorstenairy — APPROVED** [review]
LGTM

**Review by thorstenairy — DISMISSED** [review]

**Review by thorstenairy — APPROVED** [review]

**Pull request merged**